src/io/flutter/dart/FlutterDartAnalysisServer.java

```diff
@@ -98,11 +108,60 @@ private void sendSubscriptions() {
     return analysisService.edit_getAssists(file, offset, length);
   }
 
+  @Nullable
+  public List<FlutterWidgetProperty> getWidgetDescription(@NotNull VirtualFile file, int _offset) {
+    final String filePath = FileUtil.toSystemDependentName(file.getPath());
+    final int offset = analysisService.getOriginalOffset(file, _offset);
+
+    final CountDownLatch latch = new CountDownLatch(1);
+    final AtomicReference<List<FlutterWidgetProperty>> result = new AtomicReference<>();
+    final String id = analysisService.generateUniqueId();
+    responseConsumers.put(id, (resultObject) -> {
+      try {
+        final JsonArray propertiesObject = resultObject.getAsJsonArray("properties");
+        final ArrayList<FlutterWidgetProperty> properties = new ArrayList<>();
+        for (JsonElement propertyObject : propertiesObject) {
+          properties.add(FlutterWidgetProperty.fromJson(propertyObject.getAsJsonObject()));
+        }
+        result.set(properties);
+      }
+      catch (Throwable ignored) {
+      }
+      latch.countDown();
+    });
+
+    final JsonObject request = FlutterRequestUtilities.generateFlutterGetWidgetDescription(id, filePath, offset);
+    analysisService.sendRequest(id, request);
+
+    Uninterruptibles.awaitUninterruptibly(latch, 100, TimeUnit.MILLISECONDS);
+    return result.get();
+  }
+
+
+  @Nullable
+  public SourceChange setWidgetPropertyValue(int propertyId, FlutterWidgetPropertyValue value) {
+    final CountDownLatch latch = new CountDownLatch(1);
+    final AtomicReference<SourceChange> result = new AtomicReference<>();
+    final String id = analysisService.generateUniqueId();
+    responseConsumers.put(id, (resultObject) -> {
+      try {
+        final JsonObject propertiesObject = resultObject.getAsJsonObject("change");
+        result.set(SourceChange.fromJson(propertiesObject));
+      }
+      catch (Throwable ignored) {
+      }
+      latch.countDown();
+    });
+
+    final JsonObject request = FlutterRequestUtilities.generateFlutterSetWidgetPropertyValue(id, propertyId, value);
+    analysisService.sendRequest(id, request);
+
+    Uninterruptibles.awaitUninterruptibly(latch, 100, TimeUnit.MILLISECONDS);
+    return result.get();
+  }
+
   private void processString(String jsonString) {
-    // Perform a quick check to see if we should parse this into a json object.
-    if (jsonString.contains(FLUTTER_NOTIFICATION_OUTLINE_KEY)) {
-      processResponse(new Gson().fromJson(jsonString, JsonObject.class));
-    }
+    processResponse(new Gson().fromJson(jsonString, JsonObject.class));
```

Comment:
@devoncarew
I had to remove this optimization to skip all json that didn't contain a specific key. This seems like a bit of a hack of an optimization to be performing. As long as we make sure we are calling processResponse on a background thread it shouldn't be needed.

Reply:
We may want to keep this. With the available imports work, json parsing (from this location) was showing up on performance profiles. For large repos we're sending dozen's of megabytes of json to clients to parse. (I agree that the work shouldn't be done on the EDT; I don't remember if that was the case when we were looking at the performance profiles a few months ago).

What was it you're trying avoid? Are we going to need a more flexible filter mechanism?

Reply:
We would need a more flexible approach to filtering as otherwise we filter out all the analysis server messages about widget refactoring. Let me check if we are calling this on the EDT thread. If it isn't actually done on the EDT thread it shouldn't really be impacting user performance even it shows up in profiles as a few % of cpu usage.

Reply:
For posterity, here's the original perf issue: https://youtrack.jetbrains.com/issue/WEB-40646 (w/ some data about the perf profiles).

Reply:
Ok how about this. Currently this code does execute on the ServerResponseReaderThread which is the same thread that autocomplete will be using so I can understand how it is causing problems. What I can do is tweak FlutterDartAnalysisService to do its json parsing on a separate thread and return immediately. That way there should be less of an impact on other FlutterDartAnalysisServer consumers if I am understanding the threading model correctly.

Reply:
OK, sgtm. Longer term we may want to investigate a faster json parser (Alex was going to look at `com.fasterxml.jackson.core.JsonParser`), and plan to try and use the already parsed json from the Dart plugin instead of re-parsing it a 2nd time in the Flutter plugin. That would be less work overall and simplify any threading hijinks we'll have to do in the short term.

Reply:
switched to a slightly faster json parsing api (less reflection I think) and moved work to a background thread so we don't block the dart plugin.